src/dos/cdrom_ioctl_linux.cpp

```diff
@@ -113,4 +324,175 @@ bool CDROM_Interface_Ioctl::ReadSectorsHost([[maybe_unused]] void *buffer,
+		return;
+	}
+
+	std::string name = std::string(MixerChannelNamePrefix) +
+	                   std::to_string(device_number);
+
+	// Input buffer is used as a C-style buffer passed to an ioctl.
+	// Its size must always be 1 second of data.
+	// Stored in int16_t samples.
+	input_buffer.resize(REDBOOK_PCM_FRAMES_PER_SECOND * REDBOOK_CHANNELS);
+
+	// Output buffer uses C++ vector operations.
+	// Its max capacity will be 1 second so go ahead and reserve that.
+	// Stored in AudioFrame structs.
+	output_buffer.reserve(REDBOOK_PCM_FRAMES_PER_SECOND);
+
+	auto callback = std::bind(&CDROM_Interface_Ioctl::CdAudioCallback,
+	                          this,
+	                          std::placeholders::_1);
+	mixer_channel = MIXER_AddChannel(callback,
+	                                 REDBOOK_PCM_FRAMES_PER_SECOND,
+	                                 name.c_str(),
+	                                 {ChannelFeature::Stereo,
+	                                  ChannelFeature::DigitalAudio});
+}
+
+void CDROM_Interface_Ioctl::CdAudioCallback(const uint16_t requested_frames)
+{
+	if (frame_position == 0) {
+		if (sectors_remaining < 1) {
+			StopAudio();
+			return;
+		}
+		cdrom_read_audio cd = {};
+		cd.addr.lba         = current_sector;
+		cd.addr_format      = CDROM_LBA;
+		cd.buf = reinterpret_cast<uint8_t*>(input_buffer.data());
+
+		// Maximum this ioctl can read in a single call is 1 second (75
+		// redbook frames)
+		cd.nframes = std::min(sectors_remaining, CD_FRAMES);
+		constexpr auto PcmSamplesPerRedbookFrame = BYTES_PER_RAW_REDBOOK_FRAME /
+		                                           REDBOOK_BPS;
+		size_t samples = cd.nframes * PcmSamplesPerRedbookFrame;
+		current_sector += cd.nframes;
+		sectors_remaining -= cd.nframes;
+
+		if (ioctl(cdrom_fd, CDROMREADAUDIO, &cd) != 0) {
+			// Write out silence if something goes wrong
+			std::fill(input_buffer.begin(), input_buffer.end(), 0);
+
+#ifdef DEBUG_IOCTL
+			LOG_WARNING("CDROM_IOCTL: CdAudioCallback: CDROMREADAUDIO ioctl failed");
+#endif
+		}
+
+		output_buffer.clear();
+		for (size_t i = 0; i < samples; i += 2) {
+			const int16_t left  = host_to_le16(input_buffer[i]);
+			const int16_t right = host_to_le16(input_buffer[i + 1]);
+
+			// output_buffer.emplace_back(left, right);
+
+			// For some reason, emplace_back is much slower
+			// in debug build and causes noticible popping
+			// This makes the loop 4ms instead of 10ms on my system.
+			AudioFrame frame(left, right);
+			output_buffer.push_back(frame);
+		}
+	}
+
+	const size_t frames_available = output_buffer.size() - frame_position;
+	const size_t frames_to_write = std::min(static_cast<size_t>(requested_frames),
+	                                        frames_available);
+
+	mixer_channel->AddSamples_sfloat(frames_to_write,
+	                                 &output_buffer[frame_position][0]);
+
+	frame_position += frames_to_write;
+	if (frame_position >= output_buffer.size()) {
+		frame_position = 0;
+	}
```

Comment:
Great explanation.  I did actually do that logging while I was initially writing this and was curious why it was being called so frequently and asking for such a low amount of samples.  Now I know why.

I am a little bit concerned now about the `ioctl` call in the callback.  It was taking 3ms or so to complete (which seems bad if the entire tick is supposed to take 1ms, though the spike only happens once per second) but I haven't noticed any problems in practice.  The `input_buffer` size could be reduced.  It can be anywhere from 1 to 75 redbook frames (75 being 1 second).  I was just using the max to reduce the number of  total syscalls and reducing it probably just means smaller but more frequent latency spikes.

Ultimately though, it's reading from a slow optical drive so I don't know if there's much that can be done short of doing reads on a separate thread.  So maybe it's fine how it is if it doesn't appear to cause any problems?

Reply:
Finding the optimal number of redbook frames to read is easy enough to do, so I suggest to play around with it a bit (in release builds), and see if you can lower the call latency.

Ultimately, if this works well enough, then I'm happy for you to merge it. Yes, threading  is the best way to go about it, but I don't want to test your patience 😅 ; you've done a stellar job on this feature already! 🚀 🎉 

However, if you're _really_ interested in adding threading (entirely optional!), I'm here to help. Even then, I suggest merging this and adding threading in a secon PR. The patterns are already laid out, so it's a matter of following the `midi_mt32.cpp` or `midi_fluidsynth.cpp` implementations that use `std::thread` and our nice `RWQueue` (in `rwqueue.h`). But more on that if you ask me 😄